### ISSUE TITLE
[Environment]::CurrentDirectory is the Current Directory?

### ISSUE THREAD

**Issue by 237dmitry** [PowerShell/PowerShell#14950]
## Steps to reproduce

```powershell
PS > Set-Location /
PS > (Get-Location).Path
/
PS > [Environment]::CurrentDirectory
/home/user

```

## Expected behavior

```powershell
PS > Set-Location /
PS > (Get-Location).Path
/
PS > [Environment]::CurrentDirectory
/
```

## Actual behavior
`[Environment]::CurrentDirectory` shows not "Current Directory", rather "Start Directory"

## Environment data

```none

Name                           Value
----                           -----
PSVersion                      7.1.2
PSEdition                      Core
GitCommitId                    7.1.2
OS                             Linux 5.10.15-172.current #1 SMP PREEMPT Fri Feb 12 16:35:57 UTC 2021
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

**Comment by jborean93** [comment]
`[Environment]::CurrentDirectory` refers to the process wide working directory. This is different from the PowerShell location which is specific to the Runspace the pipeling is running in. They are not designed to impact each other because:

* There can be multiple runspaces per thread, if you are running multiple runspaces then each could be setting their own location overwiting the process wide working directory causing all sorts of race conditions
* A PowerShell location can be more than just a filesystem path, on Windows you have registry/cert/wsman/etc so setting a location to something that isn't a filesystem cannot translate to the current directory

In short you should always resolve your paths to the absolute path before calling any .NET functions. This will ensure that any relative directories are relative to PowerShell's location rather than the process wide currenct directory.

**Comment by 237dmitry** [comment]
> working directory

"Working Directory" sounds better, don't be misleading. Or it has to have its own path provider, like registry, variable, environment, temp... or not a string as type to avoid erroneous actions. So that the user, for example, can't take and recursively delete something.

**Comment by mklement0** [comment]
To add to @jborean93's comments:

* As unfortunate as this discrepancy is, it is the unavoidable price to pay for supporting multiple runspaces per process; this was first debated in #3428.

* It is a moot point, but if syncing the current dirs. ("working dirs." is just a synonym) _were_ possible, you could defensibly _always_ use the `FileSystem` provider's location, if the current PowerShell location is a _different_ provider's; that is, you could use `(Get-Location -PSProvider FileSystem).ProviderPath` - in fact, the latter is the robust way to refer to PowerShell's current _file-system_ location (directory).

* Unfortunately, it is currently cumbersome to resolve a path to a file/directory to its full, _native_ path, as needed for calling. NET methods:
  * If the path refers to an _existing_ item, the solution is simple: `(Convert-Path $relativePath)`
  * Otherwise, things get cumbersome, because `Convert-Path` doesn't work with _not-yet-existing_ items, as is typical when you pass an _output_ file name, for instance; you have to resort to `Join-Path (Get-Location -PSProvider FileSystem).ProviderPath $relativePath` or `[IO.Path]::GetFullPath($relativePath, (Get-Location -PSProvider FileSystem).ProviderPath)` or - only interactively, because it isn't robust - `"$PWD/$relativePath"`
  * Potential future improvements are:
     * Make `Convert-Path` support non-existing paths too - see #2993
     * Introduce a PowerShell-aware helper type that makes conversion to a full, native file-system path easy, along the lines of `[PSPathInfo] $relativePath` - see #14745

**Comment by 237dmitry** [comment]
> To add

This is all clear. But nevertheless, a situation is possible when the user "buys" the name "CurrentDirectory"... I have shown by example how this can happen. And then, there will be complaints - "powershell deleted all my pictures with kittens". It's actually dangerous. Why "Current"? .Net-refered and hard to change static property name, backward compatibility?

**Comment by mklement0** [comment]
From .NET's perspective, it _is_ the current directory, at the _process_ level, which is all that it knows about and needs to know about (it is PowerShell that builds on .NET, which itself knows nothing about PowerShell).

I get the potential for confusion, but from .NET's perspective the property name is perfectly appropriate.

The upshot is: In PowerShell:

* Don't use `[Environment]::CurrentDirectory`  - use `Get-Location` / `$PWD`
* Always use full, native paths when referring to file-system items in .NET method calls - obtaining such paths should be made easier.

The best we can do is document this.

**Comment by mklement0** [comment]
I've created two documentation issues:

* for `about_Locations`: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7318
* for `about_Methods`: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/7319

**Comment by 237dmitry** [comment]
> I get the potential for confusion

Thank you. I thought I was misunderstood with all these online translators.
